### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by yoavst**
When I tried (and failed) to convert Fuel to a multi-platform project, I've noticed that Fuel codebase needs a refactoring for two goals: 
* Make the code more Kotlin (primary constructors instead of `lateinit`s, use new methods from 1.1 instead of calling Java...)
* Hide Java implementation details out of the public interface

This refactoring attempt tries to hide the Java impl details when it is possible to do it without losing features. However the library is still Java library, and more step has to be taken to make it possible to use Fuel in multi platforms.

### CODE REVIEW

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt [hunk 1]
```diff
@@ -65,23 +66,15 @@ class Encoding : Fuel.RequestConvertible {
...
-
-    companion object {
-        private fun defaultHeaders(): MutableMap<String, Any> = mutableMapOf("Accept-Encoding" to "compress;q=0.5, gzip;q=1.0")
-    }
+    private fun queryFromParameters(params: List<Pair<String, Any?>>?): String = params.orEmpty()
+            .filterNot { it.second == null }
+            .mapTo(mutableListOf()) { "${it.first}=${it.second}" }
+            .joinToString("&")
 
+    private fun defaultHeaders(): MutableMap<String, Any> = mutableMapOf("Accept-Encoding" to "compress;q=0.5, gzip;q=1.0")
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Having return as `MutableMap` might not be such a great idea. Can we use just `Map` and let the client convert to `mutable` counterpart themselves?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt [hunk 2]
```diff
@@ -65,23 +66,15 @@ class Encoding : Fuel.RequestConvertible {
...
-                    .joinToString("&")
-        } ?: ""
-    }
-
-    companion object {
-        private fun defaultHeaders(): MutableMap<String, Any> = mutableMapOf("Accept-Encoding" to "compress;q=0.5, gzip;q=1.0")
-    }
+    private fun queryFromParameters(params: List<Pair<String, Any?>>?): String = params.orEmpty()
+            .filterNot { it.second == null }
+            .mapTo(mutableListOf()) { "${it.first}=${it.second}" }
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Maybe using just `map` is better.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/FuelManager.kt [hunk 3]
```diff
@@ -11,14 +11,9 @@ import java.security.KeyStore
 import java.util.concurrent.Executor
 import java.util.concurrent.ExecutorService
 import java.util.concurrent.Executors
-import javax.net.ssl.HostnameVerifier
-import javax.net.ssl.HttpsURLConnection
-import javax.net.ssl.SSLContext
-import javax.net.ssl.SSLSocketFactory
-import javax.net.ssl.TrustManagerFactory
+import javax.net.ssl.*
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/FuelManager.kt]: I prefer to do individual import though. What do you think?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadTaskRequest.kt [hunk 4]
```diff
@@ -3,87 +3,90 @@ package com.github.kittinunf.fuel.core.requests
...
+private const val BUFFER_SIZE = 1024
+private val CRLF = "\r\n".toByteArray()
+
+private fun guessContentType(name: String): String = try {
+    URLConnection.guessContentTypeFromName(name) ?: "application/octet-stream"
+} catch (ex: NoClassDefFoundError) {
+    // The MimetypesFileTypeMap class doesn't exists on old Android devices.
+    "application/octet-stream"
+}
+
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadTaskRequest.kt]: Can you please remove empty lines at the end?

**Comment by kittinunf** [comment]
On the first glance, this looks pretty awesome! Thanks, @yoavst for doing this.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt (hunk 2 above)
Comment (yoavst) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Have no idea why I used it like that. Fixing.

**Comment by yoavst** [comment]
@kittinunf I've fixed #220 , tell me if you are OK with the fix or I'll revert and merge.

**Comment by yoavst** [comment]
@kittinunf?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt [hunk 5]
```diff
@@ -1,47 +1,48 @@
...
         var modifiedPath = path
         var data: String? = null
-        val headerPairs: MutableMap<String, Any> = defaultHeaders()
-        if (encodeParameterInUrl(method)) {
-            var querySign = ""
-            val queryParamString = queryFromParameters(parameters)
-            if (queryParamString.isNotEmpty()) {
-                if (path.count() > 0) {
-                    querySign = if (path.last() == '?') "" else "?"
+        val headerPairs: MutableMap<String, Any> = defaultHeaders.toMutableMap()
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Shall we drop the explicit type declaration?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt [hunk 6]
```diff
@@ -65,23 +66,14 @@ class Encoding : Fuel.RequestConvertible {
...
-    }
-
-    companion object {
-        private fun defaultHeaders(): MutableMap<String, Any> = mutableMapOf("Accept-Encoding" to "compress;q=0.5, gzip;q=1.0")
-    }
+    private fun queryFromParameters(params: List<Pair<String, Any?>>?): String = params.orEmpty()
+            .filterNot { it.second == null }
+            .joinToString("&") { "${it.first}=${it.second}" }
 
+    private val defaultHeaders: Map<String, Any> = mapOf("Accept-Encoding" to "compress;q=0.5, gzip;q=1.0")
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Here as well.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Request.kt [hunk 7]
```diff
@@ -20,66 +20,64 @@ import java.util.concurrent.Future
...
+        var type: Type = Type.REQUEST,
+        val headers: MutableMap<String, String> = mutableMapOf(),
+        val parameters: List<Pair<String, Any?>> = listOf(),
+        var name: String = "",
+        val names: MutableList<String> = mutableListOf(),
+        val mediaTypes: MutableList<String> = mutableListOf(),
+        var timeoutInMillisecond: Int = 15000,
+        var timeoutReadInMillisecond: Int = timeoutInMillisecond) : Fuel.RequestConvertible {
+
+    @Deprecated(replaceWith = ReplaceWith("method"), message = "http naming is deprecated, use 'method' instead")
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Request.kt]: 👏

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 8]
```diff
@@ -1,11 +1,7 @@
 package com.github.kittinunf.fuel.toolbox
 
 import com.github.kittinunf.fuel.Fuel
-import com.github.kittinunf.fuel.core.Client
-import com.github.kittinunf.fuel.core.FuelError
-import com.github.kittinunf.fuel.core.Method
-import com.github.kittinunf.fuel.core.Request
-import com.github.kittinunf.fuel.core.Response
+import com.github.kittinunf.fuel.core.*
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: I think we should perfer the import to be expanded.

**Comment by kittinunf** [comment]
I have final check on this and commented with some small nits. After you fixed them, I think we could merge.


Thanks @yoavst for your hardwork!!! 👏 ✨

**Comment by yoavst** [comment]
Back from the trip on the 31th, will do it then, using the new debugger duck I bought 😀

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/FuelError.kt [hunk 9]
```diff
@@ -1,11 +1,6 @@
...
 
-import kotlin.properties.Delegates
-
-class FuelError : Exception() {
-    var exception: Exception by Delegates.notNull()
-    var errorData = ByteArray(0)
-    var response = Response()
-
+class FuelError(val exception: Exception, val errorData: ByteArray = ByteArray(0), val response: Response = Response.error())
+    : Exception("FuelError", exception) {
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/FuelError.kt]: We should propagate the error so the info could be able to extract by clients. I suggest we use this version of the constructor `class FuelError(..) : Exception(exception)` instead.

File: build.gradle [hunk 10]
```diff
@@ -21,7 +21,7 @@ allprojects {
         publish_version = '1.9.0'
 
         //dependencies version
-        kotlin_version = '1.1.3'
+        kotlin_version = '1.1.4'
```
Comment (kittinunf) [inline build.gradle]: Do you think we can just jump to version 1.1.4-2 altogether?

**Comment by yoavst** [comment]
@kittinunf is it OK?

**Comment by kittinunf** [comment]
👍 👍 👍 merge as soon as you are ready

**Pull request merged**